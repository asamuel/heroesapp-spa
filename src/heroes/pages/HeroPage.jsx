import { Navigate, useNavigate, useParams } from "react-router-dom";
import { useMemo } from "react";
import { getHeroById } from "../helpers/getHeroById";

export const HeroPage = () => {
  const { id } = useParams();
  const navigate = useNavigate();
  const hero = useMemo(() => getHeroById(id), [id]);
  const heroUrl = `/heroes/${id}.jpg`;

  const onNavigateBack = () => {
    navigate(-1);
  };
  if (!hero) return <Navigate to={"/"} />;
  return (
    <div className="row mt-5">
      <div className="col-4 animate__animated animate__zoomIn">
        <img src={heroUrl} alt={hero.superhero} className="img-thumbnail" />
      </div>

      <div className="col-8">
        <h3>{hero.superhero}</h3>
        <ul className="list-group list-group-flush">
          <li className="list-group-item">
            <b>Alterego: </b> {hero.alter_ego}
          </li>
          <li className="list-group-item">
            <b>Publisher: </b> {hero.publisher}
          </li>
          <li className="list-group-item">
            <b>First Appearance: </b> {hero.first_appearance}
          </li>
        </ul>

        <h5 className="mt-3">Characters</h5>
        <p>{hero.characters}</p>

        <button className="btn btn-outline-primary" onClick={onNavigateBack}>
          Back
        </button>
      </div>
    </div>
  );
};
