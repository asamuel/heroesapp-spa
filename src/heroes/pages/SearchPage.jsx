import { useLocation, useNavigate } from "react-router-dom";
import { useForm } from "../hooks/useForm";
import { useMemo } from "react";
import queryString from "query-string";
import { HeroItem } from "../components/HeroItem";
import { getHeroesByName } from "../helpers/getHeroesByName";

export const SearchPage = () => {
  const navigate = useNavigate();
  const location = useLocation();
  const { q = "" } = queryString.parse(location.search);
  const heroes = useMemo(() => getHeroesByName(q), [q]);
  const showSearch = q.length === 0;
  const showError = q.length > 0 && heroes.length === 0;
  const { searchValue, onInputChange } = useForm({ searchValue: q });

  const onSubmitSearch = (event) => {
    event.preventDefault();
    navigate(`?q=${searchValue}`);
  };
  return (
    <div className="row">
      <div className="col-5">
        <h3>Searh</h3>
        <hr />

        <form onSubmit={onSubmitSearch} aria-label="form">
          <input
            type="text"
            name="searchValue"
            value={searchValue}
            onChange={onInputChange}
            placeholder="search..."
            className="form-control"
            autoComplete="off"
          />
        </form>
      </div>

      <div className="col-7">
        <h3>Results</h3>
        <hr />

        <div
          className="alert alert-primary animate__animated animate__fadeIn"
          style={{ display: showSearch ? "" : "none" }}
        >
          Searh a Hero
        </div>
        <div
          aria-label="alert"
          className="alert alert-danger animate__animated animate__fadeIn"
          style={{ display: showError ? "" : "none" }}
        >
          No hero with <b>{q}</b>
        </div>

        {heroes.map((x) => (
          <HeroItem key={x.id} {...x} />
        ))}
      </div>
    </div>
  );
};
