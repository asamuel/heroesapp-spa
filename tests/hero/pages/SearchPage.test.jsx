import { fireEvent, render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import { SearchPage } from "../../../src/heroes/pages/SearchPage";

const mockedUseNavigate = jest.fn();

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useNavigate: () => mockedUseNavigate,
}));

describe("test on SearchPage", () => {
  beforeEach(() => jest.clearAllMocks());

  test("should match snapshot", () => {
    const { container } = render(
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  test("should input contain batman", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=batman"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const inputSearch = screen.getByRole("textbox");
    const imgBat = screen.getByRole("img");
    const alert = screen.getByLabelText("alert");

    expect(inputSearch.value).toBe("batman");
    expect(imgBat.src).toContain("batman.jpg");
    expect(alert.style.display).toBe("none");
  });

  test("should show error if error not exists", () => {
    render(
      <MemoryRouter initialEntries={["/search?q=kakaroto"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const alert = screen.getByLabelText("alert");

    expect(alert.style.display).toBeFalsy();
  });

  test("should call navigate when form was submit", () => {
    render(
      <MemoryRouter initialEntries={["/search"]}>
        <SearchPage />
      </MemoryRouter>
    );

    const input = screen.getByRole("textbox");
    const form = screen.getByRole("form");

    fireEvent.input(input, { target: { value: "iron" } });
    fireEvent.submit(form);

    expect(mockedUseNavigate).toHaveBeenCalledWith("?q=iron");
  });
});
