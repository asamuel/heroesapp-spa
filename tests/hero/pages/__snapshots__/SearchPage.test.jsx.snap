// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test on SearchPage should match snapshot 1`] = `
<div>
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h3>
        Searh
      </h3>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchValue"
          placeholder="search..."
          type="text"
          value=""
        />
      </form>
    </div>
    <div
      class="col-7"
    >
      <h3>
        Results
      </h3>
      <hr />
      <div
        class="alert alert-primary animate__animated animate__fadeIn"
      >
        Searh a Hero
      </div>
      <div
        aria-label="alert"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No hero with 
        <b />
      </div>
    </div>
  </div>
</div>
`;
